src/compas/CompasSaveTo.ts

```diff
@@ -0,0 +1,174 @@
+import {customElement, html, LitElement, property, TemplateResult} from "lit-element";
+import {InfoDetail, newLogEvent, newWizardEvent, Wizard, WizardInput} from "../foundation.js";
+import {get, translate} from "lit-translate";
+
+import {TextFieldBase} from "@material/mwc-textfield/mwc-textfield-base";
+import {OpenSCD} from "../open-scd.js";
+import {CompasChangeSetRadiogroup} from "./CompasChangeSet.js";
+import {CompasScltypeRadiogroup} from "./CompasScltypeRadiogroup.js";
+import {addSclDocument, updateSclDocument} from "./CompasService.js";
+
+import '../compas/CompasChangeSet.js';
+import './CompasScltypeRadiogroup.js';
+
+@customElement('compas-save-to')
+export class CompasSaveTo extends LitElement {
+  @property({type: String})
+  docName!: string;
+  @property({type: String})
+  docId!: string;
+
+  getNameField() : TextFieldBase {
+    return <TextFieldBase>this.shadowRoot!.querySelector('mwc-textfield[id="name"]');
+  }
+
+  getSclTypeRadioGroup() : CompasScltypeRadiogroup {
+    return (<CompasScltypeRadiogroup>this.shadowRoot!
+      .querySelector("compas-scltype-radiogroup"))
+  }
+
+  getChangeSetRadiogroup(): CompasChangeSetRadiogroup {
+    return (<CompasChangeSetRadiogroup>this.shadowRoot!
+      .querySelector("compas-changeset-radiogroup"))
+  }
+
+  checkValidity(): boolean {
+    if (!this.docId) {
+      return this.getNameField().checkValidity()
+        && this.getSclTypeRadioGroup().checkValidity();
+    }
+    return this.getChangeSetRadiogroup().checkValidity();
+  }
+
+  render(): TemplateResult {
+    return this.renderWizardPage();
+  }
+
+  renderWizardPage(): TemplateResult {
+    if (!this.docId) {
+      return html`
+        <mwc-textfield dialogInitialFocus id="name" label="${translate('scl.name')}"
+                       value="${this.docName}" required>
+        </mwc-textfield>
+
+        <compas-scltype-radiogroup .value="${getTypeFromDocName(this.docName)}"></compas-scltype-radiogroup>
+      `;
+    }
+    return html `
+      <compas-changeset-radiogroup></compas-changeset-radiogroup>
+    `;
+  }
+}
+
+function getTypeFromDocName(docName: string) {
+  if (docName!.lastIndexOf(".") == docName!.length - 4) {
+    return docName.substring(docName.lastIndexOf(".") + 1);
+  }
+  throw "Unable to determine type from document name!";
+}
+
+function addSclToCompass(wizard: Element, compasSaveTo: CompasSaveTo, doc: XMLDocument) {
+  const openScd = <OpenSCD>document.querySelector('open-scd');
+  let name = compasSaveTo.getNameField()!.value;
+  if (name!.lastIndexOf(".") == name!.length - 4) {
```

Comment:
I feel like '!' is used a lot in here.
If it's necessary everywhere (telling the compiler the expression cannot be null/undefined), then it's fine. But for me it feels like sanity checking.

Reply:
Or is this just Typescript default? (Not very Typescript expert! :D )

Reply:
Removed the ones that are not needed for the compiler.